# Homework Assignment - Lesson 6: Combining Datasets - Joins
**Name:** [Alexander Weis]

**Due Date:** [10/12/2025]

**Instructions:**
- Complete the following tasks in this R notebook
- Use the exact variable names specified for each task
- Ensure your code runs without errors and produces the expected outputs
- Submit your completed notebook with all code and outputs

---

## Part 1: Data Import and Setup

**Task:** Import the following CSV files and examine their structure.

**Required Variable Names:**
You must use these EXACT variable names:
- `customers` for customers.csv
- `orders` for orders.csv  
- `order_items` for order_items.csv
- `products` for products.csv
- `suppliers` for suppliers.csv

**Expected Outputs:**
- All datasets loaded successfully
- Print dataset dimensions using `nrow()` and `ncol()`
- Show first 3 rows of each dataset using `head(dataset, 3)`

In [1]:
# Your code here:
# 1. Load the tidyverse package
# 2. Set working directory to sample_datasets  
# 3. Import all CSV files using read_csv() with EXACT variable names above
# 4. Print dimensions and head() for each dataset

# Load tidyverse
library(tidyverse)

cat("📚 Loaded tidyverse package successfully!\n")
cat("🔗 Join functions now available: inner_join(), left_join(), right_join(), full_join()\n")
cat("🎯 Ready to learn about combining datasets!\n")
# Set working directory
setwd("/workspaces/assignment-1-Alex-Weis4-2/data")
# Import datasets with exact variable names
customers <- read_csv("customers.csv")
orders <- read_csv("orders.csv")
order_items <- read_csv("order_items.csv")
products <- read_csv("products.csv")
suppliers <- read_csv("suppliers.csv")

# Print dataset information (uncomment after creating variables above):
cat("Dataset Dimensions:\n")
cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
cat("Order Items:", nrow(order_items), "rows x", ncol(order_items), "columns\n")
cat("Products:", nrow(products), "rows x", ncol(products), "columns\n")
cat("Suppliers:", nrow(suppliers), "rows x", ncol(suppliers), "columns\n")

# Show first 3 rows of each dataset (uncomment after loading data):
# head(customers, 3)
# head(orders, 3)
# head(order_items, 3)
# head(products, 3)
# head(suppliers, 3)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.2     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


📚 Loaded tidyverse package successfully!
🔗 Join functions now available: inner_join(), left_join(), right_join(), full_join()
🎯 Ready to learn about combining datasets!


Rows: 100 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Name, Email, City
dbl  (1): CustomerID
date (1): Registration_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 250 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl  (3): OrderID, CustomerID, Total_Amount
date (1): Order_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 400 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (4): OrderID, ProductID, Quantity, Unit_Price

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this m

Dataset Dimensions:
Customers: 100 rows x 5 columns
Orders: 250 rows x 4 columns
Order Items: 400 rows x 4 columns
Products: 50 rows x 4 columns
Suppliers: 10 rows x 3 columns


## Part 2: Basic Joins

**Tasks:**
1. **Inner Join:** Create `customer_orders_inner` by joining customers and orders
2. **Left Join:** Create `customer_orders_left` to show all customers  
3. **Right Join:** Create `customer_orders_right` to show all orders
4. **Full Join:** Create `customer_orders_full` for complete data view

Use the exact variable names specified above for autograding. Analyze row counts and explain results.

In [11]:
# Your code here:
# Create an inner join between customers and orders
# Analyze the results and compare row counts
customer_orders <- inner_join(customers, orders, by = "CustomerID")
print("Inner Join (matching CustomerIDs only):")
print(customer_orders)

[1] "Inner Join (matching CustomerIDs only):"
# A tibble: 200 × 8
   CustomerID Name       Email        City  Registration_Date OrderID Order_Date
        <dbl> <chr>      <chr>        <chr> <date>              <dbl> <date>    
 1          1 Customer 1 customer1@e… Phoe… 2020-10-03             87 2023-03-28
 2          1 Customer 1 customer1@e… Phoe… 2020-10-03            214 2023-09-12
 3          2 Customer 2 customer2@e… Los … 2020-06-02            173 2024-02-25
 4          2 Customer 2 customer2@e… Los … 2020-06-02            190 2023-04-19
 5          3 Customer 3 customer3@e… Chic… 2021-04-20             29 2023-03-07
 6          3 Customer 3 customer3@e… Chic… 2021-04-20            146 2023-04-10
 7          4 Customer 4 customer4@e… Hous… 2022-06-16              7 2023-05-01
 8          4 Customer 4 customer4@e… Hous… 2022-06-16             61 2024-01-19
 9          5 Customer 5 customer5@e… Phoe… 2023-08-22             15 2023-12-27
10          5 Customer 5 customer5@e… Phoe…

In [15]:
# Your code here:
# Create a left join to keep all customers
# Count how many customers have no orders
customer_orders_left <- left_join(customers, orders, by = "CustomerID")
no_orders_count <- sum(is.na(customer_orders_left$OrderID))
print(paste("Left Join (all customers, including those without orders):", nrow(left_joined_data), "rows"))
print(paste("Number of customers without orders:", no_orders_count))


[1] "Left Join (all customers, including those without orders): 200 rows"


[1] "Number of customers without orders: 0"


In [16]:
# Your code here:
# Create a right join to keep all orders
# Check for orders with invalid customer IDs
customer_orders_right <- right_join(customers, orders, by = "CustomerID")
invalid_customer_orders <- sum(is.na(customer_orders_right$CustomerName))
print(paste("Right Join (all orders, including those with invalid customer IDs):", nrow(right_joined_data), "rows"))
print(paste("Number of orders with invalid customer IDs:", invalid_customer_orders))


Warning message:
“Unknown or uninitialised column: `CustomerName`.”


[1] "Right Join (all orders, including those with invalid customer IDs): 250 rows"
[1] "Number of orders with invalid customer IDs: 0"


In [86]:
<VSCode.Cell id="#VSC-0da36c1e" language="markdown">
## Part 2: Basic Joins

**Required Variable Names:**
You must create these EXACT variable names:

**Part 2.1:** Create `customer_orders` using inner_join()
**Part 2.2:** Create `customer_orders_left` using left_join()  
**Part 2.3:** Create `customer_orders_right` using right_join()
**Part 2.4:** Create `customer_orders_full` using full_join()

**Expected Outputs:**
For each join, print:
- Number of rows in the result using `nrow()`
- Comparison with original dataset sizes
- Analysis of unmatched records

ERROR: Error in parse(text = input): <text>:1:1: unexpected '<'
1: <
    ^


In [13]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:
customer_orders <- inner_join(customers, orders, by = "CustomerID")
print("Inner Join (matching CustomerIDs only):")
print(customer_orders)

# Required output for autograding:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

[1] "Inner Join (matching CustomerIDs only):"
# A tibble: 200 × 8
   CustomerID Name       Email        City  Registration_Date OrderID Order_Date
        <dbl> <chr>      <chr>        <chr> <date>              <dbl> <date>    
 1          1 Customer 1 customer1@e… Phoe… 2020-10-03             87 2023-03-28
 2          1 Customer 1 customer1@e… Phoe… 2020-10-03            214 2023-09-12
 3          2 Customer 2 customer2@e… Los … 2020-06-02            173 2024-02-25
 4          2 Customer 2 customer2@e… Los … 2020-06-02            190 2023-04-19
 5          3 Customer 3 customer3@e… Chic… 2021-04-20             29 2023-03-07
 6          3 Customer 3 customer3@e… Chic… 2021-04-20            146 2023-04-10
 7          4 Customer 4 customer4@e… Hous… 2022-06-16              7 2023-05-01
 8          4 Customer 4 customer4@e… Hous… 2022-06-16             61 2024-01-19
 9          5 Customer 5 customer5@e… Phoe… 2023-08-22             15 2023-12-27
10          5 Customer 5 customer5@e… Phoe…

In [20]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:
customer_orders_left <- left_join(customers, orders, by = "CustomerID")

# Required output for autograding:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [21]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:
customer_orders <- inner_join(customers, orders, by = "CustomerID") 

# Expected output:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [25]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:
customer_orders_left <- left_join(customers, orders, by = "CustomerID")

# Expected output:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [24]:
# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
# Your code here:
customer_orders_right <- right_join(customers, orders, by = "CustomerID")

# Expected output:
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:
")
cat("Total rows:", nrow(customer_orders_right), "
")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "
")

Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 


In [22]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:
customer_orders_full <- full_join(customers, orders, by = "CustomerID")

# Expected output:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [23]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:
customer_orders_full <- full_join(customers, orders, by = "CustomerID")

# Required output for autograding:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


## Part 3: Multi-Table Joins

**Tasks:**
1. Create a comprehensive dataset by joining `orders`, `customers`, and `order_items`
2. Extend this dataset by adding `products` and `suppliers` information
3. Create a complete supply chain view showing the full customer-to-supplier relationship

Build your joins step by step and examine the results at each stage.

## Part 3: Multi-Table Joins

**Requirements for Autograding:**
You must create these EXACT variable names in order:

1. `orders_items` - Join orders and order_items  
2. `orders_customers_items` - Add customers to above
3. `complete_order_data` - Add products information
4. `complete_data` - Add suppliers for complete supply chain view

**Expected Outputs:**
Print the number of rows for each step to show the join progression.

In [26]:
# Step 1: REQUIRED variable name: orders_items
# Your code here:
orders_items <- inner_join(orders, order_items, by = "OrderID")

# Required output for autograding:
cat("Step 1 - Orders with Items:", nrow(orders_items), "rows\n")

Step 1 - Orders with Items: 400 rows


In [28]:
# Step 2: REQUIRED variable name: orders_customers_items  
# Your code here:
orders_customers_items <- inner_join(orders_items, customers, by = "CustomerID")

# Required output for autograding:
cat("Step 2 - Add Customers:", nrow(orders_customers_items), "rows\n")

Step 2 - Add Customers: 310 rows


In [33]:
# Step 3: REQUIRED variable name: complete_order_data
# Your code here:
complete_order_data <- inner_join(orders_customers_items, products, by = "ProductID")
print(complete_order_data)
# Required output for autograding:
cat("Step 3 - Add Products:", nrow(complete_order_data), "rows\n")

# A tibble: 310 × 14
   OrderID CustomerID Order_Date Total_Amount ProductID Quantity Unit_Price
     <dbl>      <dbl> <date>            <dbl>     <dbl>    <dbl>      <dbl>
 1       1         87 2023-08-30         424.         2        3      116. 
 2       1         87 2023-08-30         424.        22        5      207. 
 3       1         87 2023-08-30         424.        26        5       61.8
 4       3         37 2024-03-19         549.        19        1      475. 
 5       7          4 2023-05-01         483.         7        2      360. 
 6       8         21 2023-12-25         615.         3        2      432. 
 7       8         21 2023-12-25         615.        21        3      217. 
 8      10        100 2023-06-08        1674.         5        2      121. 
 9      13         16 2024-01-24        1787.        10        1      361. 
10      13         16 2024-01-24        1787.        18        5      385. 
# ℹ 300 more rows
# ℹ 7 more variables: Name <chr>, Email <chr>, Ci

In [34]:
# Step 4: REQUIRED variable name: complete_data
# Your code here:
complete_data <- inner_join(complete_order_data, suppliers, by = "Supplier_ID")

# Required output for autograding:
cat("Step 4 - Complete Supply Chain:", nrow(complete_data), "rows\n")

Step 4 - Complete Supply Chain: 310 rows


## Part 4: Data Quality Analysis

**Tasks:**
1. Use `anti_join()` to find unmatched records between tables
2. Use `semi_join()` to find matched records
3. Check for duplicate keys and analyze their impact on joins
4. Identify and document any data quality issues

This analysis helps understand data integrity and potential issues in your datasets.

In [ ]:
<VSCode.Cell id="#VSC-d7f91846" language="markdown">
## Part 4: Data Quality Analysis

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customers_no_orders` - Use anti_join() to find customers with no orders
2. `orphaned_orders` - Use anti_join() to find orders without customers  
3. `products_never_ordered` - Use anti_join() to find unordered products
4. `active_customers` - Use semi_join() to find customers with orders

**Expected Outputs:**
Print the count of records for each using `nrow()`.

In [35]:
# REQUIRED variable name: customers_no_orders
# Your code here:
customers_no_orders <- anti_join(customers, orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

Customers who never placed an order: 0 


In [36]:
# REQUIRED variable name: orphaned_orders
# Your code here:
orphaned_orders <- anti_join(orders, customers, by = "CustomerID")

# Required output for autograding:
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

Orders without corresponding customers: 50 


In [37]:
# REQUIRED variable name: products_never_ordered
# Your code here:
products_never_ordered <- anti_join(products, order_items, by = "ProductID")

# Required output for autograding:
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

Products that were never ordered: 0 


In [38]:
# REQUIRED variable name: active_customers
# Your code here:
active_customers <- semi_join(customers, orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who placed at least one order: 100 


In [42]:
# Data integrity checks - count duplicate keys
# Your code here: Calculate and print duplicate counts for each dataset
duplicate_customers <- sum(duplicated(customers$CustomerID))
duplicate_orders <- sum(duplicated(orders$OrderID))
duplicate_products <- sum(duplicated(products$ProductID))
duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

# Required outputs for autograding (uncomment after creating the calculations above):
duplicate_customers <- sum(duplicated(customers$CustomerID))
duplicate_orders <- sum(duplicated(orders$OrderID))
duplicate_products <- sum(duplicated(products$ProductID))
duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

cat("Data Quality Summary:\n")
cat("Duplicate customer IDs:", duplicate_customers, "\n")
cat("Duplicate order IDs:", duplicate_orders, "\n")
cat("Duplicate product IDs:", duplicate_products, "\n")
cat("Duplicate supplier IDs:", duplicate_suppliers, "\n")

Data Quality Summary:
Duplicate customer IDs: 0 
Duplicate order IDs: 0 
Duplicate product IDs: 0 
Duplicate supplier IDs: 0 


## Part 5: Business Analysis with Joined Data

**Tasks:**
1. Calculate customer lifetime value metrics (total spent, number of orders, average order value)
2. Analyze product performance (total quantity sold, revenue generated, order frequency)
3. Evaluate supplier performance (total revenue, number of products supplied)
4. Create a regional analysis by customer city

Use your joined datasets to generate meaningful business insights.

In [ ]:
<VSCode.Cell id="#VSC-f7364f43" language="markdown">
## Part 5: Business Analysis with Joined Data

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customer_metrics` - Customer analysis with Total_Spent, Order_Count, Avg_Order_Value
2. `product_metrics` - Product analysis with Total_Revenue, Total_Quantity, Order_Frequency  
3. `supplier_metrics` - Supplier analysis with Product_Count by supplier
4. `regional_analysis` - Sales analysis by customer city

**Required Columns:**
- customer_metrics: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
- product_metrics: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
- supplier_metrics: Supplier_ID, Supplier_Name, Product_Count
- regional_analysis: City, Total_Sales, Customer_Count, Avg_Customer_Value

In [50]:
# REQUIRED variable name: customer_metrics
# Your code here: Calculate customer lifetime value metrics
# Required columns: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
customer_metrics <- customers %>%
  left_join(orders, by = "CustomerID") %>%
  group_by(CustomerID, Name) %>%
  summarise(
    Total_Spent = sum(Total_Amount, na.rm = TRUE),
    Order_Count = n(),
    Avg_Order_Value = mean(Total_Amount, na.rm = TRUE),
    .groups = "drop"
  )

# Required output for autograding:
cat("Customer Analysis Summary:
")
cat("Total customers analyzed:", nrow(customer_metrics), "
")
cat("Top customer total spent: $", round(max(customer_metrics$Total_Spent, na.rm = TRUE), 2), "
")
print(head(customer_metrics[order(-customer_metrics$Total_Spent), ], 3))

Customer Analysis Summary:
Total customers analyzed: 100 
Top customer total spent: $ 3729.79 
# A tibble: 3 × 5
  CustomerID Name        Total_Spent Order_Count Avg_Order_Value
       <dbl> <chr>             <dbl>       <int>           <dbl>
1         97 Customer 97       3730.           2           1865.
2         63 Customer 63       3721.           2           1861.
3         67 Customer 67       3604.           2           1802.


In [52]:
# REQUIRED variable name: product_metrics  
# Your code here: Analyze product performance
# Required columns: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
product_metrics <- products %>%
  left_join(order_items, by = "ProductID") %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Unit_Price * Quantity, na.rm = TRUE),
    Total_Quantity = sum(Quantity, na.rm = TRUE),
    Order_Frequency = n(),
    .groups = "drop"
  )

# Required output for autograding:
cat("Product Analysis Summary:
")
cat("Total products analyzed:", nrow(product_metrics), "
")
cat("Top product revenue: $", round(max(product_metrics$Total_Revenue, na.rm = TRUE), 2), "
")
print(head(product_metrics[order(-product_metrics$Total_Revenue), ], 3))

Product Analysis Summary:
Total products analyzed: 50 
Top product revenue: $ 13970.66 
# A tibble: 3 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        26 Product 26          13971.             60              19
2        47 Product 47          12582.             38              10
3        43 Product 43          10584.             36              10


In [53]:
# REQUIRED variable name: supplier_metrics
# Your code here: Evaluate supplier performance  
# Required columns: Supplier_ID, Supplier_Name, Product_Count
supplier_metrics <- suppliers %>%
  left_join(products, by = c("Supplier_ID" = "Supplier_ID")) %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Product_Count = n(),
    .groups = "drop"
  )

# Required output for autograding:
cat("Supplier Analysis Summary:
")
cat("Total suppliers analyzed:", nrow(supplier_metrics), "
")
cat("Max products per supplier:", max(supplier_metrics$Product_Count, na.rm = TRUE), "
")
print(supplier_metrics[order(-supplier_metrics$Product_Count), ])

Supplier Analysis Summary:
Total suppliers analyzed: 10 
Max products per supplier: 10 
# A tibble: 10 × 3
   Supplier_ID Supplier_Name Product_Count
         <dbl> <chr>                 <int>
 1           7 Supplier 7               10
 2           4 Supplier 4                6
 3          10 Supplier 10               6
 4           1 Supplier 1                5
 5           3 Supplier 3                5
 6           5 Supplier 5                5
 7           6 Supplier 6                5
 8           2 Supplier 2                3
 9           8 Supplier 8                3
10           9 Supplier 9                2


In [70]:
# REQUIRED variable name: regional_analysis
# Your code here: Create regional analysis by customer city
# Required columns: City, Total_Sales, Customer_Count, Avg_Customer_Value
regional_analysis <- customers %>%
  left_join(orders, by = "CustomerID") %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Total_Amount, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Avg_Customer_Value = mean(Total_Amount, na.rm = TRUE),
    .groups = "drop"
  )

# Required output for autograding:
cat("Regional Analysis Summary:
")
cat("Total cities analyzed:", nrow(regional_analysis), "
")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "
")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 5 
Highest city sales: $ 51724.77 
# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          51725.             24              1078.
2 Los Angeles      47916.             22              1089.
3 Houston          43093.             19              1134.
4 Phoenix          38317.             19              1008.
5 New York         29941.             16               936.


In [88]:
# Your code here:
# Analyze product performance metrics
product_performance <- order_items %>%
  left_join(products, by = "ProductID") %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Quantity_Sold = sum(Quantity, na.rm = TRUE),
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Times_Ordered = n(),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))
  print(product_performance)


# A tibble: 50 × 5
   ProductID Product_Name Total_Quantity_Sold Total_Revenue Times_Ordered
       <dbl> <chr>                      <dbl>         <dbl>         <int>
 1        26 Product 26                    60        13971.            19
 2        47 Product 47                    38        12582.            10
 3        43 Product 43                    36        10584.            10
 4        11 Product 11                    36        10186.            12
 5        32 Product 32                    35        10105.            11
 6        18 Product 18                    25        10019.             6
 7        21 Product 21                    30         9725.             8
 8        15 Product 15                    32         9674.             8
 9        13 Product 13                    31         9470.             9
10        17 Product 17                    33         8926.             9
# ℹ 40 more rows


In [ ]:
# Your code here:
# Evaluate supplier performance
supplier_performance <- products %>%
  left_join(suppliers, by = "Supplier_ID") %>%
  left_join(order_items, by = "ProductID") %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Products_Supplied = n_distinct(ProductID),
    Total_Units_Sold = sum(Quantity, na.rm = TRUE),
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))


In [67]:
# Your code here:
# Create regional analysis by customer city
regional_analysis <- customers %>%
  left_join(orders, by = "CustomerID") %>%
  group_by(City) %>%
  summarise(
    Customer_Count = n_distinct(CustomerID),
    Total_Orders = sum(!is.na(OrderID)),
    Total_Revenue = sum(Total_Amount, na.rm = TRUE),
    Avg_Order_Value = mean(Total_Amount, na.rm = TRUE),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))


In [76]:
# Supplier Performance Analysis
supplier_metrics <- complete_data %>%  # could not use complete_supply_chain 
  group_by(City) %>%
  filter(!is.na(Supplier_Name)) %>%
  group_by(Supplier_ID, Supplier_Name, Country) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))

cat("Supplier Performance:\n")
print(supplier_metrics)

Supplier Performance:
# A tibble: 10 × 6
   Supplier_ID Supplier_Name Country     Total_Revenue Products_Supplied
         <dbl> <chr>         <chr>               <dbl>             <int>
 1           5 Supplier 5    Japan              43434.                 5
 2           7 Supplier 7    Germany            35416.                10
 3          10 Supplier 10   Japan              31875.                 6
 4           1 Supplier 1    USA                29706.                 5
 5           3 Supplier 3    Japan              25989.                 5
 6           4 Supplier 4    South Korea        25341.                 6
 7           6 Supplier 6    Japan              21003.                 5
 8           9 Supplier 9    Japan              14845.                 2
 9           2 Supplier 2    USA                10263.                 3
10           8 Supplier 8    Germany             6805.                 3
# ℹ 1 more variable: Orders_Involved <int>


In [77]:
# Regional Analysis by Customer City
regional_analysis <- complete_data %>% # could not use complete_supply_chain 
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Sales))

cat("Regional Sales Analysis:\n")
print(regional_analysis)

Regional Sales Analysis:
# A tibble: 5 × 5
  City        Total_Sales Customer_Count Order_Count Avg_Customer_Value
  <chr>             <dbl>          <int>       <int>              <dbl>
1 Chicago          72278.             23          36              3143.
2 Phoenix          49333.             19          32              2596.
3 Los Angeles      48506.             18          32              2695.
4 Houston          38143.             18          30              2119.
5 New York         36417.             16          26              2276.


## Part 6: Complex Business Questions

**Tasks:**
Answer the following business questions using your joined datasets:

1. **Customer Segmentation:** Identify your top 10% of customers by total spending
2. **Product Recommendations:** Which products are frequently bought together?
3. **Supplier Dependency:** Which suppliers are most critical to the business?
4. **Market Expansion:** Which cities have high customer counts but low average order values?

Provide data-driven answers with supporting analysis.

## Part 6: Complex Business Questions

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `top_customers` - Top 10% of customers by spending (use quantile function)
2. `product_combinations` - Products frequently bought together  
3. `critical_suppliers` - Suppliers ranked by importance
4. `market_expansion` - Cities with expansion opportunities

**Expected Outputs:**
Each analysis should include specific metrics and rankings as shown in the required outputs below.

In [87]:
# REQUIRED variable name: top_customers
# Your code here: Identify top 10% of customers by total spending
# Use quantile() function to find the 90th percentile threshold
top_customers <- customer_metrics %>%
  filter(Total_Spent >= quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE))
print(top_customers)
# Required output for autograding (uncomment after completing your analysis):
top_10_percent_threshold <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)
cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_10_percent_threshold, 2), "\n")
cat("Number of top customers:", nrow(top_customers), "\n")

# A tibble: 10 × 5
   CustomerID Name        Total_Spent Order_Count Avg_Order_Value
        <dbl> <chr>             <dbl>       <int>           <dbl>
 1          6 Customer 6        3364            2           1682 
 2         22 Customer 22       3230.           2           1615.
 3         30 Customer 30       3246.           2           1623.
 4         39 Customer 39       3406.           2           1703.
 5         45 Customer 45       3218.           2           1609.
 6         63 Customer 63       3721.           2           1861.
 7         65 Customer 65       3470.           2           1735.
 8         67 Customer 67       3604.           2           1802.
 9         88 Customer 88       3401.           2           1700.
10         97 Customer 97       3730.           2           1865.
Customer Segmentation Results:
Top 10% spending threshold: $ 3169.56 
Number of top customers: 10 


In [82]:
# REQUIRED variable name: product_combinations
# Your code here: Find products frequently bought together
# Hint: Group by OrderID, filter orders with multiple items
product_combinations <- order_items %>%
group_by(OrderID) %>%
 filter(n() > 1) %>%
 summarise(Products = paste(sort(ProductID), collapse = ", ")) %>%
 ungroup() %>%
 group_by(Products) %>%
 summarise(n = n(), .groups = "drop") %>%
 arrange(desc(n))

# Required output for autograding (uncomment after completing your analysis):
cat("Product Combination Analysis:\n")
cat("Total product combinations found:", nrow(product_combinations), "\n")
if(nrow(product_combinations) > 0) {
   cat("Most frequent combination appears", max(product_combinations$n, na.rm = TRUE), "times\n")
}

Product Combination Analysis:
Total product combinations found: 119 
Most frequent combination appears 2 times


In [83]:
# REQUIRED variable name: critical_suppliers
# Your code here: Analyze which suppliers are most critical
# Consider both revenue contribution and product diversity
critical_suppliers <- supplier_metrics %>%
filter(Total_Revenue >= quantile(supplier_metrics$Total_Revenue, 0.9, na.rm = TRUE) &
  Products_Supplied >= quantile(supplier_metrics$Products_Supplied, 0.9, na.rm = TRUE))

# Required output for autograding (uncomment after completing your analysis):
cat("Critical Suppliers Analysis:\n")
cat("Total suppliers analyzed:", nrow(critical_suppliers), "\n")
print("Top 3 most critical suppliers:")
print(head(critical_suppliers, 3))

Critical Suppliers Analysis:
Total suppliers analyzed: 0 
[1] "Top 3 most critical suppliers:"
# A tibble: 0 × 6
# ℹ 6 variables: Supplier_ID <dbl>, Supplier_Name <chr>, Country <chr>,
#   Total_Revenue <dbl>, Products_Supplied <int>, Orders_Involved <int>


In [85]:
# REQUIRED variable name: market_expansion  
# Your code here: Identify market expansion opportunities
# Focus on cities with multiple customers but varying order values
market_expansion <- regional_analysis %>%
filter(Customer_Count >= quantile(regional_analysis$Customer_Count, 0.75, na.rm = TRUE) &
  Avg_Customer_Value <= quantile(regional_analysis$Avg_Customer_Value, 0.25, na.rm = TRUE))

# Required output for autograding (uncomment after completing your analysis):
cat("Market Expansion Analysis:\n")
cat("Cities evaluated for expansion:", nrow(market_expansion), "\n")
print("Top expansion opportunities:")
print(head(market_expansion, 5))

Market Expansion Analysis:
Cities evaluated for expansion: 0 
[1] "Top expansion opportunities:"
# A tibble: 0 × 5
# ℹ 5 variables: City <chr>, Total_Sales <dbl>, Customer_Count <int>,
#   Order_Count <int>, Avg_Customer_Value <dbl>


## Part 7: Summary and Insights

**Requirements for Autograding:**
Complete the analysis summary below with specific findings from your work. Your responses will be evaluated for completeness and accuracy.

**Grading Criteria:**
- Specific metrics and numbers from your analysis
- Clear identification of data quality issues  
- Actionable business recommendations
- Technical insights about join performance and appropriateness

## Analysis Summary

### Key Findings from Join Operations:
- **Join Efficiency:** [There are lots of matched records between customers and orders, and the other data too.  The joins had the same number of rows as the left table, so they were efficient.]
- **Data Relationships:** [There were many customers without orders, and many orders without customers, indicating potential data entry issues.  I did not see any duplicates either.]
- **Multi-table Joins:** [The data grew wider as I added more tables, but the number of rows dropped a little for the missing data.]

### Data Quality Issues Discovered:
- **Orphaned Records:** [There were 50 orphaned orders without matching customers.]
- **Missing Data:** [The tables with missing or invalid references were orders and order_items.]
- **Referential Integrity:** [There were some customers without valid ID.]

### Business Insights:
- **Top Customer:** [The highest spending customer was customer 97 and spent $3730 and made 2 orders.]
- **Best Product:** [Product 26 made 13971 in revenue.]
- **Regional Performance:** [It looks like Chicago has the highest number of sales and revenue.]
- **Supplier Analysis:** [Supplier 7 provided the most products at 10.]

### Strategic Recommendations:
- [I would spread out marketing to focus on more cities, since Chicago is doing well but other cities are not.  I would also try to reach out to customers who have not ordered yet, and for my last recommendation I would try to diversify ordering products from other countries and or supplier since supplier 7 since they seem to be doing a lot of that.  The company could get into trouble if supplier 7 went under.]

### Technical Learnings:
- **Inner vs Left Joins:** [Inner join is more efficient when you only need matched records, while left join is useful for retaining all records from the primary table.]
- **Multi-table Strategy:** [I found it easier to build joins step by step, verifying each stage before proceeding.]
- **Performance Considerations:** [The joins were very efficient since it combined them and the data set went down in size.]

In [ ]:
# Optional: Save your analysis results (not required for autograding)
# Your code here: Save key datasets to CSV files if desired


# Final summary output for autograding verification:
cat("\n=== HOMEWORK 6 COMPLETION SUMMARY ===\n")
cat("✓ Data Import: 5 datasets loaded\n")
cat("✓ Basic Joins: 4 join types completed\n") 
cat("✓ Multi-table Joins: 4-step progression completed\n")
cat("✓ Data Quality: Anti-joins and semi-joins performed\n")
cat("✓ Business Analysis: Customer, product, supplier, and regional analysis completed\n")
cat("✓ Complex Questions: Advanced business scenarios analyzed\n")
cat("✓ Summary: Comprehensive insights documented\n")
cat("\nAll required variables created for autograding verification.\n")